lib/exp.js

```diff
@@ -0,0 +1,2655 @@
+ *
+ * @return (integer value)
+ */
+exports.intRscan = _paramsExp(exp.ops.INT_RSCAN)
+
+/**
+ * Create expression that returns the minimum value in a variable number of expressions.
+ * All arguments must be the same type (integer or float).
+ * Requires server version 5.6.0+.
+ *
+ * @param ... Variable number of integer or float expressions.
+ * @return (integer or float value)
+ */
+exports.min = _VAExp(exp.ops.MIN)
+
+/**
+ * Create expression that returns the maximum value in a variable number of expressions.
+ * All arguments must be the same type (integer or float).
+ * Requires server version 5.6.0+.
+ *
+ * @param ... Variable number of integer or float expressions.
+ * @return (integer or float value)
+ */
+exports.max = _VAExp(exp.ops.MAX)
+
+/*********************************************************************************
+ * FLOW CONTROL AND VARIABLE EXPRESSIONS
+ *********************************************************************************/
+/**
+ * Conditionally select an expression from a variable number of expression pairs
+ * followed by default expression action. Requires server version 5.6.0+.
+ *
+ * @return first action expression where bool expression is true or action-default.
+ */
+exports.cond = _VAExp(exp.ops.COND)
+
+/**
+ * Define variables and expressions in scope.
+ * Requires server version 5.6.0+.
+ *
+ * @param ... Variable number of expression def followed by a scoped
+ *  expression.
+ * @return result of scoped expression.
+ */
+exports.let = _VAExp(exp.ops.LET)
+
+/**
+ * Assign variable to an expression that can be accessed later.
+ * Requires server version 5.6.0+.
+ *
+ * @param varName Variable name.
+ * @param expr The variable is set to the result of expr.
+ * @return A variable name expression pair.
+ */
+exports.def = (varName, expr) => [
+  ..._rawStr(varName),
+  ...expr
+]
+
+/**
+ * Retrieve expression value from a variable.
+ * Requires server version 5.6.0+.
+ *
+ * @param varName Variable name.
+ * @return value stored in variable.
+ */
+exports.var = (varName) => [
+  { op: exp.ops.VAR, count: 2 },
+  ..._rawStr(varName)
+]
+
+/*********************************************************************************
+ * LIST READ EXPRESSIONS
+ *********************************************************************************/
+
+function getListType (type, returnType, isMulti) {
+  let expected = type
+
+  switch (returnType & ~lists.returnType.INVERTED) {
+    case lists.returnType.INDEX:
+    case lists.returnType.REVERSE_INDEX:
+    case lists.returnType.RANK:
+    case lists.returnType.REVERSE_RANK:
+      expected = isMulti ? exp.type.LIST : exp.type.INT
+      break
+    case lists.returnType.COUNT:
+      expected = exp.type.INT
+      break
+    case lists.returnType.VALUE:
+      if (isMulti) {
+        expected = exp.type.LIST
+      }
+      break
+    case lists.returnType.NONE:
+    default:
+      console.log('getListType valueType:' + type + 'returnType:' + returnType + ' FAILED')
```

Comment:
Instead of logging and returning an `ERROR` type, maybe the client should just raise a `TypeError` here. What do you think?

Reply:
sure will add that